x-pack/plugins/enterprise_search/public/applications/shared/tables/inline_editable_table/inline_editable_table_logic.ts

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { kea, MakeLogicType } from 'kea';
+
+import { FormErrors, InlineEditableTableColumn, ItemWithAnID } from './types';
+
+interface InlineEditableTableActions<Item extends ItemWithAnID> {
+  deleteItem(item: Item): { item: Item };
+  doneEditing(): void;
+  editNewItem(): void;
+  editExistingItem(item: Item): { item: Item };
+  reorderItems(items: Item[], oldItems: Item[]): { items: Item[]; oldItems: Item[] };
+  saveExistingItem(): void;
+  saveNewItem(): void;
+  setEditingItemValue(newValue: Item): { item: Item };
+  setFormErrors(formErrors: FormErrors): { formErrors: FormErrors };
+}
+
+const generateEmptyItem = <Item extends ItemWithAnID>(
+  columns: Array<InlineEditableTableColumn<Item>>
+): Item => {
+  const emptyItem = columns.reduce((acc, column) => ({ ...acc, [column.field]: '' }), {}) as Item;
+  return emptyItem;
+};
+
+const getUnsavedItemId = () => null;
+const doesIdMatchUnsavedId = (idToCheck: number) => idToCheck === getUnsavedItemId();
+
+interface InlineEditableTableValues<Item extends ItemWithAnID> {
+  // TODO This could likely be a selector
+  isEditing: boolean;
+  // TODO we should editingItemValue have editingItemValue and editingItemId should be a selector
+  editingItemId: Item['id'] | null; // editingItem is null when the user is editing a new but not saved item
+  editingItemValue: Item | null;
+  formErrors: FormErrors;
+  isEditingUnsavedItem: boolean;
+  doesEditingItemValueContainEmptyProperty: boolean;
+}
+
+interface InlineEditableTableProps<Item extends ItemWithAnID> {
+  columns: Array<InlineEditableTableColumn<Item>>;
+  instanceId: string;
+  // TODO Because these callbacks are params, they are only set on the logic once (i.e., they are cached)
+  // which makes using "useState" to back this really hard.
+  onAdd(item: Item, onSuccess: () => void): void;
+  onDelete(item: Item, onSuccess: () => void): void;
+  onReorder?(items: Item[], oldItems: Item[], onSuccess: () => void): void;
+  onUpdate(item: Item, onSuccess: () => void): void;
+  transformItem?(item: Item): Item;
+  validateItem?(item: Item): FormErrors;
+}
+
+type InlineEditableTableLogicType<Item extends ItemWithAnID> = MakeLogicType<
+  InlineEditableTableValues<Item>,
+  InlineEditableTableActions<Item>,
+  InlineEditableTableProps<Item>
+>;
+
+export const InlineEditableTableLogic = kea<InlineEditableTableLogicType<ItemWithAnID>>({
+  path: (key: string) => ['enterprise_search', 'inline_editable_table_logic', key],
```

Comment:
I remember losing so much time before realizing I had to use a function for the path

Reply:
I remember Byron swooping in like an eagle and solving this for me! 🎉